src/puppetlabs/puppetdb/utils.clj

```diff
@@ -343,6 +360,29 @@
     {:command command :version (Integer. version)
      :certname certname :payload (json/->RawJsonString payload)}))
 
+(def ^Charset utf8 StandardCharsets/UTF_8)
+
+(defn utf8-bytes [s] (.getBytes ^String s utf8))
+
+(defn utf8-length
+  "Return the length in bytes of the given string when encoded in UTF-8"
+  [s]
+  (count (.getBytes ^String s utf8)))
+
+(defn utf8-truncate
+  "Truncate the given string such that its UTF-8 representation is at most
+  `max-bytes` long. Note that the returned string may be empty."
+  [s max-bytes]
+  (let [utf8-bytebuff (ByteBuffer/allocate max-bytes)
```

Comment:
@senior hmm, i see what you mean, and so far it's not completely clear to me from the docs that the bytebuffer from 0 to position is *guaranteed* to be a valid encoding when flush returns OVERFLOW, but I suspect that for at least an encoding like UTF-8, it will be.